search/api.py

```diff
@@ -0,0 +1,261 @@
+"""
+Functions for search
+"""
+from itertools import islice
+import logging
+
+from django.conf import settings
+from django.contrib.auth.models import User
+from elasticsearch.helpers import bulk
+from elasticsearch.exceptions import NotFoundError
+from elasticsearch_dsl import Mapping
+from elasticsearch_dsl.connections import connections
+
+from profiles.models import Profile
+from profiles.serializers import ProfileSerializer
+from search.exceptions import ReindexException
+
+log = logging.getLogger(__name__)
+
+
+USER_DOC_TYPE = 'user'
+DOC_TYPES = (USER_DOC_TYPE, )
+_CONN = None
+# When we create the connection, check to make sure all appropriate mappings exist
+_CONN_VERIFIED = False
+
+
+def get_conn(verify=True):
+    """
+    Lazily create the connection.
+    """
+    # pylint: disable=global-statement
+    global _CONN
+    global _CONN_VERIFIED
+
+    do_verify = False
+    if _CONN is None:
+        _CONN = connections.create_connection(hosts=[settings.ELASTICSEARCH_URL])
+        # Verify connection on first connect if verify=True.
+        do_verify = verify
+
+    if verify and not _CONN_VERIFIED:
+        # If we have a connection but haven't verified before, do it now.
+        do_verify = True
+
+    if not do_verify:
+        if not verify:
+            # We only skip verification if we're reindexing or
+            # deleting the index. Make sure we verify next time we connect.
+            _CONN_VERIFIED = False
+        return _CONN
+
+    # Make sure everything exists.
+    index_name = settings.ELASTICSEARCH_INDEX
+    if not _CONN.indices.exists(index_name):
+        raise ReindexException("Unable to find index {index_name}".format(
+            index_name=index_name
+        ))
+
+    mappings = _CONN.indices.get_mapping()[index_name]["mappings"]
+    for doc_type in DOC_TYPES:
+        if doc_type not in mappings.keys():
+            raise ReindexException("Mapping {doc_type} not found".format(
+                doc_type=doc_type
+            ))
+
+    _CONN_VERIFIED = True
+    return _CONN
+
+
+def _index_users_chunk(users):
+    """
+    Add/update a small number of user records in Elasticsearch
+
+    Args:
+        users (list of User):
+            List of users
+
+    Returns:
+        int: Number of items inserted into Elasticsearch
+    """
+
+    conn = get_conn()
+    insert_count, errors = bulk(
+        conn,
+        (serialize_user(user) for user in users),
```

Comment:
In generators parentheses are required: https://docs.python.org/3/reference/expressions.html#generator-expressions